WIP Iterated Integrals

deeepeshthakur opened the pull request:
Implementation of Stochastic Iterated Integrals approximations from [SDELab](https://doi.org/10.1016/j.cam.2006.05.037)

ChrisRackauckas commented:
You mean SDELab2?

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,130 @@
+
+    if typef(dW) <: Number || is_diagonal_noise(integrator.sol.prob)
+        cache.WikJ = 1//2 .* J .^ 2
+    else
```
what about the commutative noise case? That should be in the RKMilCommute cache constructor already.

deeepeshthakur replied:
We don't have a function for commutative noise like we have for diagonal so I thought let's consider it in general case. Otherwise, It would be difficult maintaining a commutative noise version of every solver.

ChrisRackauckas replied:
That's what this is: https://github.com/JuliaDiffEq/StochasticDiffEq.jl/blob/master/src/perform_step/low_order.jl#L198-L202 . It should just call to a common routine in here to fill `WikJ` instead, and if it's commutative that's the calculation.

deeepeshthakur commented:
> You mean SDELab2?

SDELab2 had a few mistakes, so I tried sticking to the paper.

ChrisRackauckas commented:
Oh okay. Was just checking for licensing reasons. If it's from the paper description then you're good.

deeepeshthakur commented:
IIRC SDELAB2 is also MIT license😋

deeepeshthakur commented:
@ChrisRackauckas 
Right now what I've done is that I've added an additional method `RKMil_General` that can handle 1-D, Diagonal, Commutative, and General M-Dimensional noise. I'm testing it. If it passes the tests then we can replace the other `RKMil` methods with this in a different PR.
Also for commutative noise, I've added an option in alg part.

devmotion commented on src/iterated_integrals.jl:
```diff
@@ -15,9 +15,6 @@ function get_iterated_I!(integrator, cache::StochasticDiffEqConstantCache)
         m_seq  = cache.m_seq
         sum_dW² = zero(eltype(dW))
```
That line is not needed it seems?

deeepeshthakur replied:
Yes. Missed that. Thank You

ChrisRackauckas commented on src/perform_step/low_order.jl:
```diff
@@ -229,3 +229,124 @@ end
+  dW = W.dW
+
+  if integrator.alg.is_commutative
+    WikJ = 1//2 .* vec(dW) .* vec(dW)'
```
I think this could be handled as a branch in the iterated I just like diagonal?

deeepeshthakur replied:
But there we have a function for diagonal noise and the commutative noise it's alg specific

ChrisRackauckas replied:
The way to generate WikJ for commutative noise isn't alg-specific.

deeepeshthakur replied:
No not that. I mean we don't have an efficient way to check for commutative noise like we have for diagonal noise so rn commutative noise is just an option in the alg struct

ChrisRackauckas replied:
I see. Make it a value type so that we can dispatch on it and do a cleaner WikJ construction, and we should find a better system for noise classifications later.

ChrisRackauckas commented on src/perform_step/low_order.jl:
```diff
@@ -229,3 +229,124 @@ end
+  if integrator.alg.is_commutative
+    WikJ = 1//2 .* vec(dW) .* vec(dW)'
+  else
+    WikJ = get_iterated_I!(integrator, cache)
```
We should probably have a struct for WikJDiagonal, WikJCommute, WikJGeneral that dispatches to different ways of filling it? Then it could have the appropriate cache size for each case, and the cache variables won't necessarily need to be in the method cache (allowing for simultaneous different WikJ's in future methods)

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,237 @@
+    WikJCommute_iip{typeof(WikJ)}(WikJ)
+end
+
+function fill_WikJGeneral_oop(ΔW)
```
fill_WikJCommute_oop?

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,237 @@
+    end
+    vec_ζ = false .* vec(ΔW)
+    vec_η = similar(vec_ζ)
+    Gp1 = zeros(M)
```
Doing this because you want the lower triangle only? It's probably better to just make the matrix so that this can be GPU-friendly.

deeepeshthakur replied:
Yes. So You are saying that I should just make this a normal matrix instead?

ChrisRackauckas replied:
let's keep that for a future PR. This is complex already.

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,237 @@
+        Wik.WikJ = 1//2 .* dW .^ 2
+    else
+        @.. WikJ = 1//2*dW^2
+        @.. Wik.WikJ = WikJ
```
this line isn't needed?

deeepeshthakur replied:
I forgot that. Sorry.

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,237 @@
+function get_iterated_I!(dW, Wik::WikJDiagonal_iip)
+    @unpack WikJ = Wik
+    if typeof(dW) <: Number
+        Wik.WikJ = 1//2 .* dW .^ 2
```
this works whether dW is a number or an array?

deeepeshthakur replied:
It was not working for a number so I made the struct mutable and added this branching.

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,237 @@
+    mul!(WikJ,vec(dW),vec(dW)')
+    @.. WikJ *= 1//2
+
+    @.. Wik.WikJ = WikJ
```
this isn't needed? WikJ === Wik.WikJ

ChrisRackauckas commented on src/perform_step/low_order.jl:
```diff
@@ -229,3 +229,127 @@ end
+    if typeof(dW) <: Number || is_diagonal_noise(integrator.sol.prob)
+      WikJ = WikJ .- 1//2 .* dt
+    else
+      WikJ -= 1//2 .* dt .* Eye{eltype(dW)}(length(dW))
```
Are you sure about the identity matrix here?

deeepeshthakur replied:
It was working for the normal RKMil_Commute, so I don't see why it should be a problem here.

ChrisRackauckas replied:
oh yes, now I remember why it's there.

ChrisRackauckas commented on src/perform_step/low_order.jl:
```diff
@@ -229,3 +229,127 @@ end
+  integrator.f(du₁,uprev,p,t)
+  integrator.g(L,uprev,p,t)
+  @.. mil_correction = zero(eltype(u))
+  ggprime_norm = zero(eltype(ggprime))
```
just do `local ggprime_norm`

deeepeshthakur replied:
I'm not sure if I'm getting what you are saying here.

ChrisRackauckas replied:
instead of initializing it you can just localize it

ChrisRackauckas commented on test/sde/sde_convergence_tests.jl:
```diff
@@ -22,6 +22,8 @@ sim  = test_convergence(dts,prob,LambaEM(),trajectories=Int(1e2))
 @test abs(sim.𝒪est[:l2]-.5) < 0.2
```
we need a non-commutative convergence test now. Does https://epubs.siam.org/doi/abs/10.1137/09076636X have one? IIRC it does

deeepeshthakur replied:
Yes, It does. with m = 4. We should also find something similar for Stratonovich problems

ChrisRackauckas commented:
If you use RKMil_General and RKMil on the same random seed do they compute exactly the same values, with the same timestepping, when it's diagonal noise? Same with RKMilCommute? I think both of those algorithms can be subsumed, or at least the former can be renamed (for testing) and this can be made `RKMil`. From a quick look, it seems that the caching is done correctly so it will be just as efficient in that case as well, and the error estimator should be exactly the same.

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,237 @@
+    @.. Wik.WikJ = WikJ
+    return nothing
+end
+
```
Comment or docstring that points to where this algorithm comes from.

ChrisRackauckas replied:
In fact, describe the computation that you did, like https://github.com/JuliaDiffEq/DiffEqBase.jl/blob/master/src/nlsolve/newton.jl#L1-L43

deeepeshthakur replied:
This sounds nice. Because it's quite complicated even though I tried to be as close to the paper as I could.

deeepeshthakur replied:
I've added some markdown style docs. Hope this makes it easier to understand.

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,237 @@
+end
+
+function get_iterated_I!(dW, Wik::WikJCommute_oop)
+    @unpack WikJ = Wik
```
why have a cached version at all then? This Wik type should probably have no fields

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,237 @@
+        k += 1
+      end
+    end
+    WikJCommute_oop{eltype(ΔW), typeof(WikJ)}(WikJ, m_seq)
```
```suggestion
    WikJGeneral_oop{eltype(ΔW), typeof(WikJ)}(WikJ, m_seq)
```

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,237 @@
+
+    WikJ *= 1//2
+    a2ₚ = (π^2)/6
+    p = Int(floor((1/π)*sqrt(M/(24*dt))*sqrt(m + 4*sum_dW²/dt) + 1))
```
ceil? Also, this should have a tweakable parameter `C` from the algorithm

deeepeshthakur replied:
This I have taken from SDELAB2, There C is taken to be 1

deeepeshthakur replied:
I did `floor(ans + 1)`. Will change it to ceil.

ChrisRackauckas replied:
he was probably being lazy with C. Make C=1 the default but it should still be a parameter.

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,237 @@
+    p = Int(floor((1/π)*sqrt(M/(24*dt))*sqrt(m + 4*sum_dW²/dt) + 1))
+    for i in 1:p
+        a2ₚ -= (1/i^2)
+        var = sqrt(dt/(2*π*i))
```
zeta has a sqrt(dt), eta has a sqrt(dt) but then in Ap you divide by dt, so this should be 1/sqrt(dt)?

ChrisRackauckas replied:
oh nevermind, then it's multiplied by dt... that is written horribly in the paper lol

deeepeshthakur replied:
Yes. I skipped some scalar calculations because they canceled themselves.

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,237 @@
+        vec_ζ = randn(m)*var
+        vec_η = randn(m)*var
+        WikJ += (vec_ζ*vec_η' - vec_η*vec_ζ')
+        Aₚ -= (2/sqrt(π*i))*vec_ζ
```
little a?

deeepeshthakur replied:
I changed this part to vector calculations ![image](https://user-images.githubusercontent.com/26359940/62458031-989fce80-b799-11e9-9bd0-83891bfa289d.png). So `A_p` is the vector of `a_i0`.

deeepeshthakur replied:
And I think the `i` in line 174 is r in the paper.

ChrisRackauckas replied:
can we change the names so it lines up?

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,237 @@
+    end
+
+    WikJ -= 1//2*(dW*Aₚ' - Aₚ*dW')
+    WikJ += (sqrt(a2ₚ)*dt/π)*WikJ2
```
why is there a dt here?

deeepeshthakur replied:
![image](https://user-images.githubusercontent.com/26359940/62457606-d3553700-b798-11e9-9ddc-1178255a3ee2.png)
This is the scalar `Delta_t`/`2*pi`*`ap^0.5`. But I skipped the division by 2 due to cancellation

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,237 @@
+        WikJ2[m_seq[i,2], m_seq[i,1]] = -Gp[i]
+    end
+
+    WikJ *= 1//2
```
for legibility, this stuff should come before the tail correction

deeepeshthakur replied:
I did this because of a memory optimization.

ChrisRackauckas replied:
how does it optimize the memory?

deeepeshthakur replied:
There We'll need one extra WikJ type array if we do that.

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,237 @@
+    end
+
+    WikJ *= 1//2
+    a2ₚ = (π^2)/6
```
I don't see this in the paper at all.

deeepeshthakur replied:
Sum(1/x^2) from 1 to inf is pi^2/6. And then I subtract 1/x^2 from 1 to p.

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,237 @@
+    WikJ = dW*dW'
+    Gp1 = randn(M)
+    α = sqrt(1 + sum_dW²/dt)
+    Gp2 = Gp1/(sqrt(2)*(1+α)*dt)
```
I fail to see how this is Sigma infty?

deeepeshthakur replied:
No. Actually, I've changed the order in which we multiply scalars. Sigma infty is `a_p`

ChrisRackauckas replied:
can you rename a_p then?

deeepeshthakur replied:
Oh sorry, you are referring to the operator sigma infinity. I broke that down into some smaller operators.

ChrisRackauckas commented:
Pulling in @YingboMa as well since this is a quite important one.

deeepeshthakur commented:
> If you use RKMil_General and RKMil on the same random seed do they compute exactly the same values, with the same timestepping, when it's diagonal noise? Same with RKMilCommute? I think both of those algorithms can be subsumed, or at least the former can be renamed (for testing) and this can be made `RKMil`. From a quick look, it seems that the caching is done correctly so it will be just as efficient in that case as well, and the error estimator should be exactly the same.

I was thinking that this was the whole purpose of this. I mean I thought we would completely remove the other 3 methods when this gets going and rename `RKMil_General` to `RKMil`

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,286 @@
+    Aᵢ::Vector{eltype(rateNoiseElTypeNoUnits)}
+end
+
+function fill_WikJDiagonal_oop(ΔW)
```
This should just be the constructor for the type, i.e. have the same name as the type.

ChrisRackauckas commented on src/caches/basic_method_caches.jl:
```diff
@@ -116,3 +116,49 @@ function alg_cache(alg::RKMilCommute,prob,u,ΔW,ΔZ,p,rate_prototype,noise_rate_
+function alg_cache(alg::RKMil_General,prob,u,ΔW,ΔZ,p,rate_prototype,noise_rate_prototype,uEltypeNoUnits,uBottomEltypeNoUnits,tTypeNoUnits,uprev,f,t,dt,::Type{Val{false}})
+  if typeof(ΔW) <: Number || is_diagonal_noise(prob)
+    WikJ = fill_WikJDiagonal_oop(ΔW)
+  elseif alg.is_commutative
```
since there is a type change here, this is something we want to resolve at compile time.

deeepeshthakur replied:
Then we'll have to write different functions for cache initializations.

ChrisRackauckas replied:
no you won't

deeepeshthakur replied:
How can we resolve this in compile-time without making separate functions?

ChrisRackauckas replied:
make it type-based, like `Val(true)` or something.

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,286 @@
+"""
+
+    get_iterated_I!(dW, Wik::AbstractWikJ, C)
```
this needs to be right above `get_iterated_I!` for the docstring to work.

ChrisRackauckas commented on src/iterated_integrals.jl:
````diff
@@ -0,0 +1,286 @@
+```
+
+```math
+    Oper1(Gₚ) = (√(Σ∞)/2 Gₚ) = (Iₘ/√2 + 1/(√2 * (1 + α) * Δt) Kₘ(Iₘ² - Pₘ)(Iₘ ⨂ W(Δt)W(Δt)ᵀ)(Iₘ² - Pₘ)Kₘᵀ)Gₚ
````
If this is Gp1, then call it Gp1. If it's not Gp1, what is it?

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,286 @@
+    for i in 1:M
+        Gp2[i] = WikJ2[m_seq[i,1], m_seq[i,2]]
+    end
+    Gp = Gp/sqrt(2) + Gp2
```
Gp on the right hand side here is undefined? Also, where does this relation come from?

deeepeshthakur replied:
![image](https://user-images.githubusercontent.com/26359940/62858818-0196c180-bd19-11e9-9e24-9a31d48ecc8d.png)

So what I'm actually doing is applying the Square root sigma infinity operator on `Gp`. So this can be broken down into `I_M` part and the remaining part.
So when I'm saying ` Gp1 = Gp1/sqrt(2) + Gp2` here `Gp1` is the initial `Gp1` as `I_M * Gp1` will still be `Gp` and `Gp2` is the rest of the operator applied to `Gp1`. Thus this expression will give the Sigma_Infinity Operator of `Gp1`.

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,286 @@
+    for i in 1:M
+        Gp2[i] = WikJ2[m_seq[i,1], m_seq[i,2]]
+    end
+    Gp1 = Gp1/sqrt(2) + Gp2
```
@..

deeepeshthakur commented on src/caches/basic_method_caches.jl:
```diff
@@ -116,3 +116,54 @@ function alg_cache(alg::RKMilCommute,prob,u,ΔW,ΔZ,p,rate_prototype,noise_rate_
+
+function alg_cache(alg::RKMil_General,prob,u,ΔW,ΔZ,p,rate_prototype,noise_rate_prototype,uEltypeNoUnits,uBottomEltypeNoUnits,tTypeNoUnits,uprev,f,t,dt,::Type{Val{false}})
+  m = (typeof(ΔW) <: Number || is_diagonal_noise(prob))*1 + (alg.is_commutative)*2 + (!(typeof(ΔW) <: Number || is_diagonal_noise(prob) || alg.is_commutative))*3
+  WikJ = fill_WikJ(ΔW,Val(m),Val(false))
```
@ChrisRackauckas I tried this. But I don't see how this might have any compile-time benefits?

ChrisRackauckas commented on src/caches/basic_method_caches.jl:
```diff
@@ -155,15 +159,18 @@ function alg_cache(alg::RKMil_General,prob,u,ΔW,ΔZ,p,rate_prototype,noise_rate
+
+  if typeof(ΔW) <: Number || is_diagonal_noise(prob)
+    # WikJ = WikJDiagonal_iip(ΔW)
+    WikJ = fill_WikJ(ΔW,Val(1),Val(true))
```
?

deeepeshthakur replied:
I still don't understand the purpose of Value types here. 
Earlier what I did was
```
  m = (typeof(ΔW) <: Number || is_diagonal_noise(prob))*1 + (alg.is_commutative)*2 + (!(typeof(ΔW) <: Number || is_diagonal_noise(prob) || alg.is_commutative))*3
  WikJ = fill_WikJ(ΔW,Val(m),Val(false))
```
Got compilation error.
Then I changed it to:
```
    WikJ = WikJDiagonal_oop(ΔW,Val(true))
```
But I guess this won't make any difference because there is only 1 definition of `WikJDiagonal_oop`

So I did
```
    WikJ = fill_WikJ(ΔW,Val(1),Val(false))
```
1, 2, or 3 for Diagonal, commutative and general noise and true or false for in place and out of place.

ChrisRackauckas replied:
no, the purpose was for `is_commutative` to be something that is a value type so it can be a compile-time check. Examples: http://www.stochasticlifestyle.com/type-dispatch-design-post-object-oriented-programming-julia/

deeepeshthakur replied:
But then we'll have to take it as an argument. Like we can default it to diagonal noise case but then for general noise, the user will have to give another argument.

ChrisRackauckas commented on src/iterated_integrals.jl:
````diff
@@ -0,0 +1,302 @@
+```
+
+```math
+    Oper2(Gp₁) = (√(Σ∞)/2 Gp₁) = (Iₘ/√2 + F/(√2 * (1 + α) * Δt))(Gp₁)
````
what's Oper2?

deeepeshthakur replied:
I just named it so that I won't have to write the whole left part again.

ChrisRackauckas replied:
But what is it? What operation is it? Is it in the code?

deeepeshthakur replied:
No. Just explanation purposes.

ChrisRackauckas commented on src/iterated_integrals.jl:
````diff
@@ -0,0 +1,302 @@
+```
+
+"""
+
````
need to put this over the correct dispatch

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,302 @@
+
+    #operator Kₘ(Iₘ² - Pₘ)
+    WikJ2 = WikJ2 - WikJ2'
+    for i in 1:M
```
this loop isn't necessary?

ChrisRackauckas replied:
nevermind

deeepeshthakur replied:
I think we need this.

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,302 @@
+thus applying operator `F` on `Gp₂`. And hence we have
+
+    Gp₂ = F(Gp₂)
+    Gp₁ = (Gp₁/√2) + Gp₂
```
`(1-1/sqrt(2))Gp₁ = Gp₂`? That doesn't seem right.

deeepeshthakur replied:
No no. This is an assignment. This is `=` not `==`.

ChrisRackauckas replied:
It shouldn't be. The whole purpose of the docstring is to be a clear unambiguous description of what is being calculated. This is precisely the issue. You shouldn't write mathematical descriptions in a way that is dependent on the order that it is read in: that's for code which has an order.

ChrisRackauckas commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,302 @@
+
+thus applying operator `F` on `Gp₂`. And hence we have
+
+    Gp₂ = F(Gp₂)
```
F is the identity?

deeepeshthakur replied:
F is mentioned above. line 124.

ChrisRackauckas replied:
`Gp₂ = F(Gp₂)` => `F` is the identity.

deeepeshthakur replied:
This is the same issue as above. https://github.com/JuliaDiffEq/StochasticDiffEq.jl/pull/209#discussion_r314223449

ChrisRackauckas replied:
Yes it is

deeepeshthakur replied:
Fixed it. I think.

ChrisRackauckas commented:
I think the description is good now. I refactored the iterated integral setup code so that it ends up type stable from a val dispatch on whether to use the commutative noise approximation. 

I will say that this code is good to go, but it's still missing a stress test on the general noise case. We need a convergence test that showcases that the iterated integral approximation really is working. There is a test case in here which gives us good confidence, and thus I am willing to merge, but a convergence test would be better. Let's add that in a future PR.

deeepeshthakur commented:
I'll fix the undef error and typo in doc strings

devmotion commented on src/iterated_integrals.jl:
```diff
@@ -24,7 +24,7 @@ struct WikJGeneral_oop <: AbstractWikJGeneral
     function WikJGeneral_oop(ΔW)
         m = length(ΔW)
-        M = m*(m-1)/2
+        M = Int(m*(m-1)/2)
```
I'd suggest `M = div(m * (m - 1), 2)` instead. A quick check with
```julia
f(m) = div(m * (m - 1), 2); g(m) = Int(m * (m - 1) / 2)
```
shows that the output of `@code_native` and `@code_llvm` is a lot simpler for `f`.

ChrisRackauckas commented:
Note `RKMil_General` shouldn't be needed, since it could just be `RKMil`, which will auto-specialize to the special case of diagonal noise? We need to check that it's just as efficient there, but if it is, then this can be the one algorithm for the RKMils. RKMilCommute can stay an alias to the IICommutative version with a deprecation warning. That should be a followup PR.

YingboMa commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,345 @@
+end
+
+function WikJGeneral_iip(ΔW)
+    WikJ = false .* ΔW .* ΔW'
```
`false .* vec(ΔW) .* vec(ΔW)'`

YingboMa commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,345 @@
+    vec_η = false .* vec(ΔW)
+    Gp₁ = false .* Array{eltype(ΔW)}(undef, M)
+    Gp₂ = false .* Array{eltype(ΔW)}(undef, M)
+    Aᵢ = false .* vec(ΔW)
```
Those `false .*` are unnecessary and inefficient, because the desired array has the same shape with `ΔW`, so `zero` and `fill! + similar` is sufficient.

YingboMa commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,345 @@
+    @unpack WikJ, WikJ2, WikJ3, m_seq, vec_ζ, vec_η, Gp₁, Gp₂, Aᵢ = Wik
+
+    m      = length(dW)
+    M      = Int(m*(m-1)/2)
```
`div(m * (m - 1), 2)` or `m * (m - 1) ÷ 2` would be better, as @devmotion suggested.

YingboMa commented on src/iterated_integrals.jl:
```diff
@@ -0,0 +1,345 @@
+    m      = length(dW)
+    M      = Int(m*(m-1)/2)
+
+    sum_dW² = dW' * dW #zero(eltype(dW))
```
I am not too familiar with `StochasticDiffEq`, but if `dW` could be a row vector, then `dW' * dW` shouldn't be used.

```julia
julia> using LinearAlgebra

julia> dW = rand(5)';

julia> dW'dW
5×5 Array{Float64,2}:
 0.240635   0.096258   0.0756145  0.172557   0.382082
 0.096258   0.0385048  0.030247   0.0690255  0.152839
 0.0756145  0.030247   0.0237603  0.0542223  0.120061
 0.172557   0.0690255  0.0542223  0.123739   0.273987
 0.382082   0.152839   0.120061   0.273987   0.606672

julia> dot(dW, dW)
1.0333109944478744
```

ChrisRackauckas replied:
it shouldn't ever be a row vector, but dot is probably better anyways

YingboMa commented on src/perform_step/low_order.jl:
```diff
@@ -229,3 +229,127 @@ end
+            integrator.opts.reltol, integrator.opts.delta, integrator.opts.internalnorm, t)
+    else
+      En = integrator.opts.internalnorm(W.dW,t)^3*ggprime_norm^2 / 6
+      tmp = @.. integrator.opts.internalnorm(dt*(du₂ - du₁)/2,t)
```
This doesn't look right. `@.. integrator.opts.internalnorm(...)` applies the norm to each element of the array. Do you mean

```julia
tmp = integrator.opts.internalnorm((@.. dt*(du₂ - du₁)/2),t)
```
?

ChrisRackauckas replied:
it should be the latter

YingboMa commented on src/perform_step/low_order.jl:
```diff
@@ -229,3 +229,127 @@ end
+    else
+      En = integrator.opts.internalnorm(W.dW,t)^3*ggprime_norm^2 / 6
+      tmp = @.. integrator.opts.internalnorm(dt*(du₂ - du₁)/2,t)
+      tmp = @.. calculate_residuals(tmp, En, uprev, u, integrator.opts.abstol,
```
Same here.

deeepeshthakur replied:
IIRC this is the backend definition of calculate_residuals.

deeepeshthakur replied:
Nevermind.

YingboMa commented on src/perform_step/low_order.jl:
```diff
@@ -229,3 +229,127 @@ end
+    for i in 1:length(dW)
+      K = uprev + dt*du₁ + sqdt*@view(L[:,i])
+      gtmp = integrator.g(K,p,t)
+      ggprime = (gtmp - L)/sqdt
```
```suggestion
      ggprime = @.. (gtmp - L)/sqdt
```

YingboMa commented on src/perform_step/low_order.jl:
```diff
@@ -229,3 +229,127 @@ end
+  if integrator.opts.adaptive
+      En = integrator.opts.internalnorm(W.dW,t)^3*ggprime_norm^2 / 6
+      integrator.f(du₂,K,p,t+dt)
+      @.. tmp = integrator.opts.internalnorm(integrator.opts.delta * dt * (du₂ - du₁) / 2,t) + En
```
Same here.

deeepeshthakur replied:
This error estimate is taken from `RKMil_Commute` and `RKMil` So I think it should work here too.

deeepeshthakur replied:
And also this will change the type of `tmp`.

deeepeshthakur commented:
Added the suggestions except the last one.

ChrisRackauckas commented:
This still needs a real test for non-commutative order 1 convergence with at least a simple Milstein, but the iterated integrals are at least correct enough that I am going to merge it so it can at least be tracked. We should get that test working ASAP.
